lib/rucio/core/rule.py

```diff
@@ -1789,18 +2060,25 @@ def get_expired_rules(total_workers, worker_number, limit=100, blocked_rules=[],
     :param session:            Database session in use.
     """
 
-    query = session.query(models.ReplicationRule.id, models.ReplicationRule.rse_expression).filter(models.ReplicationRule.expires_at < datetime.utcnow(),
-                                                                                                   models.ReplicationRule.locked == false(),
-                                                                                                   models.ReplicationRule.child_rule_id == None).\
-        with_hint(models.ReplicationRule, "index(rules RULES_EXPIRES_AT_IDX)", 'oracle').\
-        order_by(models.ReplicationRule.expires_at)  # NOQA
-
-    query = filter_thread_work(session=session, query=query, total_threads=total_workers, thread_id=worker_number, hash_variable='name')
+    stmt = select(
+        models.ReplicationRule.id,
+        models.ReplicationRule.rse_expression
+    ).where(
+        and_(models.ReplicationRule.expires_at < datetime.utcnow(),
+             models.ReplicationRule.locked == false(),
+             models.ReplicationRule.child_rule_id == None)  # NOQA: E711
+    ).with_hint(
+        models.ReplicationRule, "index(rules RULES_EXPIRES_AT_IDX)", 'oracle'
+    ).order_by(
+        models.ReplicationRule.expires_at
+    )
+    stmt = filter_thread_work(session=session, query=stmt, total_threads=total_workers, thread_id=worker_number, hash_variable='name')
 
     if limit:
-        fetched_rules = query.limit(limit).all()
-        filtered_rules = [rule for rule in fetched_rules if rule[0] not in blocked_rules]
-        if len(fetched_rules) == limit and not filtered_rules:
+        stmt = stmt.limit(limit)
+        result = session.execute(stmt).all()
+        filtered_rules = [rule for rule in result if rule[0] not in blocked_rules]
```

Comment:
I wonder if `rule['id']` is preferable to `rule[0]`.

This comment applies throughout the file.

Reply:
To solve this, I replaced all occurences of this issue with indexing akin to this:

`filtered_rules = [rule for rule in result if rule._mapping['id'] not in blocked_rules]`

`rule['id']` was unavailable as this query does not retrieve the entire object (just models.ReplicationRule.id and models.ReplicationRule.rse_expression), so I used _mapping to index it.

Reply:
I’m afraid I do not understand your comment. Yes, the query comes with two individual columns, but I don’t see why that requires anything different.

```python
>>> stmt = select(
...         models.ReplicationRule.id,
...         models.ReplicationRule.rse_expression
...     ).where(
...         and_(models.ReplicationRule.expires_at < datetime.utcnow(),
...              models.ReplicationRule.locked == false(),
...              models.ReplicationRule.child_rule_id == null())
...     ).order_by(
...         models.ReplicationRule.expires_at
...     ).with_hint(
...         models.ReplicationRule, 'INDEX(RULES RULES_EXPIRES_AT_IDX)', 'oracle'
...     )
>>> result = session.execute(stmt).all()
>>> result[0]['id']
'53f65ded554b43c1accaf8c981f90c58'
```

Reply:
Forgive me, this console session is inaccurate. It was done using SQLAlchemy version 1.4.46 and the behaviour appears to have changed in more recent versions. However, as discussed privately, [dot notation should work](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row._mapping).

> This [`Row`](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row "sqlalchemy.engine.Row") behaves like a named tuple, in that it acts as a sequence but also supports attribute name access, e.g. `row.some_column`.

I’m having difficulty deciding what is best; please do as you prefer.

Reply:
Rewrote the indexing to use dot notation instead of `_mapping` for all occurrences